cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll

```diff
@@ -719,14 +711,19 @@ predicate ssaFlow(Node nodeFrom, Node nodeTo) {
   )
 }
 
+private predicate isArgumentOfCallable(DataFlowCall call, ArgumentNode arg) {
+  arg.argumentOf(call, _)
+}
+
+/** Holds if there is def-use or use-use flow from `pun` to `nodeTo`. */
 predicate postUpdateFlow(PostUpdateNode pun, Node nodeTo) {
-  exists(Node preUpdate, Node nFrom, boolean uncertain, SsaDefOrUse defOrUse |
+  exists(UseOrPhi use, Node preUpdate |
+    adjustForPointerArith(pun, use) and
+    useToNode(use, nodeTo) and
     preUpdate = pun.getPreUpdateNode() and
-    ssaFlowImpl(defOrUse, nFrom, nodeTo, uncertain)
-  |
-    if uncertain = true
-    then preUpdate = [nFrom, getAPriorDefinition(defOrUse)]
-    else preUpdate = nFrom
+    not exists(DataFlowCall call |
+      isArgumentOfCallable(call, preUpdate) and isArgumentOfCallable(call, nodeTo)
```

Comment:
Final question from my side. I think I mostly follow along now, but would definitely appreciate @rdmarsh2's input.

What is the reason for ignoring the argument positions here? Assuming this is important, could we add a test for this?

Reply:
It shouldn't be important, no. The main thing is that we want to disallow flow from the `PostUpdateNode` and back into the function as an argument (which violates the evaluation order). So it doesn't really matter what the argument order is since the `PostUpdateNode` always represents the value after we've returned from the function and the `ArgumentNode` always represents the value before we've entered the function.

Reply:
With that said, I'd also be interested in knowing if I've missed something here (cc @rdmarsh2).

Reply:
I think there would never be a correct step directly from the postupdate to a preupdate on the same call... That ought to only be possible in a loop, and there should be an intervening phi node in that case. If it does come up I think it's an IR inconsistency problem.

Reply:
Yeah, I agree. The closest we can come to flow directly from a `PostUpdateNode` and back to the argument is something like:
```c
int x;
// ...
while(...) {
  write_to_arg(&x);
}
```
where we'd have flow from `x`'s post update note, to a phi node at the loop entry and back to `&x`. But as Robert says, there should always be a `PhiNode` here.